src/GSoft.Extensions.MediatR.Analyzers/Internals/RoslynExtensions.cs

```diff
@@ -0,0 +1,62 @@
+﻿using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace GSoft.Extensions.MediatR.Analyzers.Internals;
+
+internal static class RoslynExtensions
+{
+    public static INamedTypeSymbol? GetBestTypeByMetadataName(this Compilation compilation, string typeMetadataName, string? assemblyName = null)
+    {
+        var assemblies = GetAllAssemblies(compilation);
+
+        if (assemblyName != null)
+        {
+            assemblies = assemblies.Where(x => x.Name == assemblyName);
+        }
+
+        foreach (var assembly in assemblies)
+        {
+            if (assembly.GetTypeByMetadataName(typeMetadataName) is { } symbol)
+            {
+                return symbol;
+            }
+        }
+
+        return null;
+    }
+
+    private static IEnumerable<IAssemblySymbol> GetAllAssemblies(Compilation compilation)
+    {
+        yield return compilation.Assembly;
+
+        foreach (var reference in compilation.References)
+        {
+            if (compilation.GetAssemblyOrModuleSymbol(reference) is IAssemblySymbol assembly)
+            {
+                yield return assembly;
+            }
+        }
+    }
+
+    public static void ReportDiagnostic(this SymbolAnalysisContext context, DiagnosticDescriptor diagnosticDescriptor, ISymbol symbol)
```

Comment:
[Nit] Should this method be named `ReportDiagnostics`, considering you are iterating over a collection?

Reply:
We still pass a single diagnostic descriptor though

Reply:
Also `Locations` is an enumerable, but in our analysis rules there's always a **single** location, not multiple

Reply:
Works for me 👍